third-party/rsutils/include/rsutils/signal.h

```diff
@@ -0,0 +1,150 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2023 Intel Corporation. All Rights Reserved.
+
+#pragma once
+
+#include "subscription.h"
+
+#include <map>
+#include <mutex>
+#include <functional>
+#include <memory>
+
+
+namespace rsutils {
+
+
+// Outside signal<> so can be easy to reference - they should be the same regardless of the type of signal
+using subscription_slot = int;
+
+
+// Signals are callbacks with multiple targets.
+// 
+// Each "slot" is a subscriber that's called when the signal is "raised."
+// 
+// There are intentionally no operators here (especially operator()!) as, despite the nice-to-look-at style, are not
+// verbose enough to be really useful.
+// 
+// The template Args are the callback arguments, and therefore what's expected to be passed to raise(). The callbacks
+// return void on purpose, for simplicity.
+// 
+// NOTE: there are more efficient implementations (e.g., with a vector). We kept it map-based for simplicity.
+//
+template< typename... Args >
+class signal
+{
+    using callback = std::function< void( Args... ) >;
+    using map = std::map< subscription_slot, callback >;
+
+    // We use a shared_ptr to control lifetime: only while it's alive can we remove subscriptions
+    struct impl
+    {
+        std::mutex mutex;
+        map subscribers;
+
+        bool remove( subscription_slot slot )
+        {
+            std::lock_guard< std::mutex > locker( mutex );
+            return subscribers.erase( slot );
+        }
+    };
+    std::shared_ptr< impl > const _impl;
+
+    signal( const signal & other ) = delete;
+    signal & operator=( const signal & ) = delete;
+
+public:
+    signal() : _impl( std::make_shared< impl >() ) {}
+    signal( signal && ) = default;
+    signal & operator=( signal && ) = default;
+
+    // Get the slot without any fancy subscription
+    subscription_slot add( callback && func )
+    {
+        std::lock_guard< std::mutex > locker( _impl->mutex );
+        // NOTE: we should maintain ordering of subscribers: later subscriptions should be called after earlier ones, so
+        // the key should keep increasing in value
+        auto slot = _impl->subscribers.empty() ? 0 : (_impl->subscribers.rbegin()->first + 1);
+        _impl->subscribers.emplace( slot, std::move( func ) );
+        return slot;
+    }
+
+    // Opposite of add()
+    bool remove( subscription_slot slot ) { return _impl->remove( slot ); }
+
+    // Unlike add(), subscribing implies a subscription, meaning the caller needs to store the return value or else it
+    // will get unsubscribed immediately!!
+    subscription subscribe( callback && func )
```

Comment:
Can you subscribe callbacks that receive arguments? `subscription` is using `defered::fn` which does not receive any

Reply:
That's the beauty of it: a subscription doesn't have to know anything about what it's unsubscribing. It doesn't even have to work with `signal` -- you can have anything return a `subscription`.

The "callback" that goes into a subscription is always without arguments -- it has to be called from a destructor. if you want to pass stuff in, you do it thru the lambda, like `signal` does it.

Reply:
OK, I get it now.
Maybe the name `subscription` is not the best, as it is not like a subscription to a publisher. It is only a one time call to a function, suitable to be used as a deleter.

Reply:
Right, a `subscription` is just a `deferred` function but with different semantics: one main difference is the `operator=()` calling `cancel()`.

Think of it like a subscription in English, not a subscription in the context of publisher-subscriber.

Reply:
I am not too sure about what are the benefits of using a `subscription` over a `deferred` function though

Reply:
The semantics.
Deferred will not `cancel()` on assignment.
Deferred has no `cancel()`.
Try to think in English verbs. `deferred` is a good utility to have. `subscription` is a use-case with its own set of requirements..